### PR TITLE
Clean up context attribute definitions

### PR DESCRIPTION
**Pull request by dapomeroy**
Creates a context object and simplifies how contexts are stored and passed.

### CODE REVIEW

File: lib/ramble/ramble/context.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            self.is_template = in_context.templates
+        if in_context.exclude:
+            self.exclude = in_context.exclude
+        if in_context.zips:
+            self.zips = in_context.zips
+        if in_context.matrices:
+            self.matrices = in_context.matrices
+
+
+def from_dict(context_name, in_dict):
```
Comment (douglasjacobsen) [inline lib/ramble/ramble/context.py:82]: I think this should either be a member function of `Context` (where it should set the context from a dictionary), i.e.:
```
new_context = Context()
new_context.from_dict(in_dict)
```

Or, the name should be more like `create_context_from_dict`.

**Review by douglasjacobsen — CHANGES_REQUESTED** [review]
This is great. Thanks for this!

One small request, feel free to do it either way I suggested.

File: lib/ramble/ramble/context.py (hunk 1 above)
Comment (dapomeroy) [reply lib/ramble/ramble/context.py:82]: Great, I've updated the name. I might be overthinking this, but my reasoning was that having this as a non-member function would ensure that updates to an existing instance go through the 'merge_context' logic. That way, a context 'from_dict' would start with a clean instance and avoid any potential conflict with the 'from_dict' method being invoked after vars have already been set and either not overwriting or not properly merging the two contexts.

File: lib/ramble/ramble/context.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+        Create a Context object, which holds context attributes.
+        """
+        self.env_variables = []
+        self.variables = syaml.syaml_dict()
+        self.internals = {}
+        self.templates = None
+        self.chained_experiments = []
+        self.modifiers = []
+        self.context_names = None
```
Comment (douglasjacobsen) [inline lib/ramble/ramble/context.py:33]: One other small nit, that I just noticed when you changed the name of the `from_dict` function.

Is there a reason for this to be plural? It seems like it should be singular instead. But, if there is a reason then we can leave it.

**Pull request merged**